lib/ecto/query/builder/select.ex

```diff
@@ -480,12 +480,17 @@ defmodule Ecto.Query.Builder.Select do
           # If merging with a schema, add the schema's query fields. This comes in handy if the user
           # is merging fields with load_in_query = false.
           # If merging with a schemaless source, do nothing so the planner can take all the fields.
-          case {old_expr, source} do
-            {{:&, _, [^binding]}, {_source, schema}} when not is_nil(schema) ->
-              Map.put(acc, binding, {new_kind, Enum.uniq(new_fields ++ schema.__schema__(:query_fields))})
+          case old_expr do
+            {:&, _, [^binding]} ->
+              source = Enum.at([query.from | query.joins], binding).source
```

Comment:
I spent a long time trying to figure out if this is legitimate. From what I can tell it is alright.

It seems like `from` is always indexed at `0` and the joins are iterated through in this way. Also I couldn't find anywhere in the planner that was adjusting the `take` indexes.